### ISSUE TITLE
NullReferenceException when trying to fake a non-virtual generic method method

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#480]
If I try to fake a non-virtual, non-generic method, I get a `FakeConfigurationException` because the method is non-virtual (as expected). But if the method is generic, FIE throws a `NullReferenceException`.

Example:

```c#
public class Foo
{
    public int GetValue()
    {
        return 0;
    }

    public T GetValue<T>()
    {
        return default(T);
    }
}
```

This throws the expected `FakeConfigurationException`:

```c#
A.CallTo(() => foo.GetValue()).Returns(42);
```

This throws a `NullReferenceException`:

```c#
A.CallTo(() => foo.GetValue<int>()).Returns(42);
```

**Comment by thomaslevesque** [comment]
The problem seems to be in `MethodInfoManager.HasSameBaseMethod`. Apparently, for a virtual generic method, `GetBaseMethod` returns the generic definition (`IsGenericMethodDefinition` is true), but not for a non-virtual generic method (in this case it returns the realized method itself). This causes `IsSameMethod` to return false, which in turn causes `GetInvokedMethod` to return null, and this case it not handled in `GetReasonForWhyMethodCanNotBeIntercepted`.

**Comment by blairconrad** [comment]
Thanks for raising this, @thomaslevesque.
At a quick pass, your fix appears to be sound.
The rest of the gang at FakeItEasy International are a little busy with other projects, and we're pushing for the 2.0 release when we do have time, so it may be a little while before we can do a proper review and roll the fix into the codebase. Don't despair at the delay, though. We are interested, and appreciate your contribution.

It's a bug in our reporting of a misuse of the framework, not a shortcoming in the framework. Also, it should fall below the 2.0 milestone work, so marking as **P3**.

**Comment by thomaslevesque** [comment]
> It's a bug in our reporting of a misuse of the framework, not a shortcoming in the framework

Yes, I agree. It's not a blocking issue at all.

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
